src/emulator/kernel/src/load_self.cpp

```diff
@@ -0,0 +1,335 @@
+        }
+
+        //assert(imports->version == 1);
+        //assert(imports->num_syms_vars == 0);
+        assert(imports->num_syms_unk == 0);
+
+        const uint32_t *const nids = Ptr<const uint32_t>(imports->func_nid_table).get(mem);
+        const Ptr<uint32_t> *const entries = Ptr<Ptr<uint32_t>>(imports->func_entry_table).get(mem);
+        if (!load_func_imports(nids, entries, imports->num_syms_funcs, mem)) {
+            return false;
+        }
+        
+        const uint32_t *const var_nids = Ptr<const uint32_t>(imports->var_nid_table).get(mem);
+        const Ptr<uint32_t> *const var_entries = Ptr<Ptr<uint32_t>>(imports->var_entry_table).get(mem);
+        if (!load_var_imports(var_nids, var_entries, imports->num_syms_vars, mem)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static bool load_func_exports(Ptr<const void> &entry_point, const uint32_t *nids, const Ptr<uint32_t> *entries, size_t count, KernelState &kernel, const MemState &mem) {
+    for (size_t i = 0; i < count; ++i) {
+        const uint32_t nid = nids[i];
+        const Ptr<uint32_t> entry = entries[i];
+
+        if (nid == NID_MODULE_START) {
+            entry_point = entry;
+            continue;
+        }
+        
+        if (nid == NID_MODULE_STOP || nid == NID_MODULE_EXIT)
+            continue;
+        
+        kernel.export_nids.emplace(nid,entry.address());
+        
+        if (LOG_EXPORTS) {
+            const char *const name = import_name(nid);
+            
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+        }
+        /*uint32_t *const stub = entry.get(mem);
+        stub[0] = 0xef000000; // svc #0 - Call our interrupt hook.
+        stub[1] = 0xe1a0f00e; // mov pc, lr - Return to the caller.
+        stub[2] = nid; // Our interrupt hook will read this.*/
+    }
+
+    return true;
+}
+
+static bool load_var_exports(Ptr<const void> &entry_point, const uint32_t *nids, const Ptr<uint32_t> *entries, size_t count, KernelState &kernel, const MemState &mem) {
+    for (size_t i = 0; i < count; ++i) {
+        const uint32_t nid = nids[i];
+        const Ptr<uint32_t> entry = entries[i];
+        
+        if(nid == NID_PROCESS_PARAM) {
+            kernel.process_param = entry;
+            LOG_DEBUG("\tNID {:#08x} (SCE_PROC_PARAMS) at {:#x}", nid, entry.address());
+            continue;
+        }
+        
+        if(nid == NID_MODULE_INFO) {
+            LOG_DEBUG("\tNID {:#08x} (NID_MODULE_INFO) at {:#x}", nid, entry.address());
+            continue;
+        }
+        
+        if(nid == 0x936c8a78) {
+            LOG_DEBUG("\tNID {:#08x} (SYSLYB) at {:#x}", nid, entry.address());
+            continue;
+        }
+        
+        if (LOG_EXPORTS) {
+            const char *const name = import_name(nid);
+            
+            LOG_DEBUG("\tNID {:#08x} ({}) at {:#x}", nid, name, entry.address());
+            
+        }
+        
+        /*if (nid == 0x935cd196) {
```

Comment:
Please add a comment explaining any commented code, why it's there, why it's commented etc.